lib/ramble/ramble/software_environments.py

```diff
@@ -0,0 +1,471 @@
+# Copyright 2022-2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+# <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+
+import llnl.util.tty as tty
+
+import ramble.repository
+import ramble.workspace
+import ramble.keywords
+import ramble.error
+import ramble.renderer
+import ramble.expander
+from ramble.namespace import namespace
+
+
+class SoftwareEnvironments(object):
+    """Class to represent a set of software environments
+
+    This class contains logic to take the dictionary representations of
+    software environments, and unify their format.
+    """
+
+    keywords = ramble.keywords.keywords
+
+    supported_confs = ['v1', 'v2']
+
+    def __init__(self, workspace):
+
+        self._raw_packages = {}
+        self._packages = {}
+        self._package_map = {}
+        self._raw_environments = {}
+        self._environments = {}
+        self._environment_map = {}
+        self._workspace = workspace
+        self.spack_dict = self._workspace.get_spack_dict().copy()
+
+        conf_type = self._detect_conf_type()
+
+        if conf_type not in self.supported_confs:
+            raise RambleSoftwareEnvironmentError(
+                f'Configuration type {conf_type} is not one of ' +
+                f'{str(self.supported_confs)}'
+            )
+
+        if conf_type == 'v1':
+            self._v1_setup()
+        elif conf_type == 'v2':
+            self._v2_setup()
+
+    def _detect_conf_type(self):
+        """Auto-detect the type of configuration provided.
+
+        Default configuration type is v2.
+
+        v1 configurations follow the format:
+
+        spack:
+          concretized: [true/false]
+          applications: {}
+          compilers: {}
+          mpi_libraries: {}
+
+        v2 configurations follow the format:
+
+        spack:
+          concretized: [true/false]
+          packages: {}
+          environments: {}
+        """
+
+        conf_type = 'v2'
+
+        if namespace.application in self.spack_dict or \
+                namespace.compilers in self.spack_dict or \
+                namespace.mpi_lib in self.spack_dict:
+            conf_type = 'v1'
+        if namespace.packages in self.spack_dict and \
+                namespace.environments in self.spack_dict:
+            conf_type = 'v2'
+
+        tty.debug(f'Detected config type of: {conf_type}')
+
+        return conf_type
+
+    def _v1_setup(self):
```

Comment:
I think pulling it out into its own object might be messy, just because it does a lot of internal stuff to the software_environments class.

Instead, I generalized the `if conf_type` block to a `getattr` block instead.

Let me know how you feel about this, and I can re-evaluate if desired.